File: lib/src/main/java/net/imoya/android/preference/fragment/TimePickerFragment.java
package net.imoya.android.preference.fragment;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TimePicker;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import net.imoya.android.fragment.BaseFragment;
import net.imoya.android.preference.R;
import net.imoya.android.preference.model.Time;
import net.imoya.android.util.Log;
import net.imoya.android.util.TimePickerHelper;

/**
 * {@link TimePicker} を表示する {@link Fragment}
 */
public class TimePickerFragment extends BaseFragment {
    private static final String ARGUMENT_TIME =
            TimePickerFragment.class.getName() + ".CurrentTime";
    private static final String ARGUMENT_IS_24_HOUR_VIEW =
            TimePickerFragment.class.getName() + ".Is24HourView";
    private static final String ARGUMENT_TITLE =
            TimePickerFragment.class.getName() + ".Title";

    private static final String TAG = "TimePickerFragment";

    private final Time time = new Time();

    private TimePicker picker = null;

    public void setIs24HourView(boolean is24HourView) {
        final Bundle original = this.getArguments();
        final Bundle arguments = (original != null) ? original : new Bundle();
        arguments.putBoolean(ARGUMENT_IS_24_HOUR_VIEW, is24HourView);
        if (original == null) {
            this.setArguments(arguments);
        }
    }

    @NonNull
    public Time getTime() {
        this.loadTime();
        return this.time;
    }

    public void setTime(@NonNull Time time) {
        final Bundle original = this.getArguments();
        final Bundle arguments = (original != null) ? original : new Bundle();
        arguments.putParcelable(ARGUMENT_TIME, time);
        if (original == null) {
            this.setArguments(arguments);
        }
    }

    public void setScreenTitle(@Nullable String title) {
        final Bundle original = this.getArguments();
        final Bundle arguments = (original != null) ? original : new Bundle();
        arguments.putString(ARGUMENT_TITLE, title);
        if (original == null) {
            this.setArguments(arguments);
        }
    }

    private void loadTime() {
        final TimePicker picker = this.picker;
        if (picker != null) {
            final TimePickerHelper pickerHelper = new TimePickerHelper(picker);
            this.time.setHour(pickerHelper.getHour());
            this.time.setMinute(pickerHelper.getMinute());
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);

        this.loadTime();
        outState.putParcelable(ARGUMENT_TIME, this.time);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        final Time time;
        if (savedInstanceState != null) {
            time = savedInstanceState.getParcelable(ARGUMENT_TIME);
        } else {
            time = this.requireArguments().getParcelable(ARGUMENT_TIME);
        }
        Log.d(TAG, "onCreate: time = " + time);

        if (time != null) {
            this.time.setHour(time.getHour());
            this.time.setMinute(time.getMinute());
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.preference_time_input_fragment, container, false);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        this.picker = null;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        this.picker = view.findViewById(R.id.time);
        new TimePickerHelper(this.picker)
                .setHourAndMinute(this.time.getHour(), this.time.getMinute());
        this.picker.setIs24HourView(this.requireArguments().getBoolean(ARGUMENT_IS_24_HOUR_VIEW));

        final String title = this.requireArguments().getString(ARGUMENT_TITLE);
        Log.d(TAG, "onViewCreated: title = " + title);
        this.setTitle(title);
    }
}

File: lib/src/main/java/net/imoya/android/preference/fragment/TimePickerFragment.kt
/*
 * Copyright (C) 2022 IceImo-P
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.imoya.android.preference.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TimePicker
import androidx.fragment.app.Fragment
import net.imoya.android.fragment.BaseFragment
import net.imoya.android.preference.PreferenceLog
import net.imoya.android.preference.R
import net.imoya.android.preference.model.Time
import net.imoya.android.util.TimePickerHelper

/**
 * [TimePicker] を表示する [Fragment]
 */
class TimePickerFragment : BaseFragment() {
    /**
     * 時刻
     */
    private val time = Time()

    /**
     * [TimePicker]
     */
    private var picker: TimePicker? = null

    /**
     * 24時間表示フラグを設定します。
     *
     * @param is24HourView 24時間表示フラグ
     */
    fun setIs24HourView(is24HourView: Boolean) {
        val original = this.arguments
        val arguments = original ?: Bundle()
        arguments.putBoolean(ARGUMENT_IS_24_HOUR_VIEW, is24HourView)
        if (original == null) {
            this.arguments = arguments
        }
    }

    /**
     * [TimePicker] の時刻を取得します。
     *
     * @return 時刻
     */
    fun getTime(): Time {
        loadTime()
        return Time(time.hour, time.minute, time.second)
    }

    /**
     * [TimePicker] の時刻を設定します。
     *
     * @param time 時刻
     */
    fun setTime(time: Time) {
        val original = this.arguments
        val arguments = original ?: Bundle()
        arguments.putParcelable(ARGUMENT_TIME, time)
        if (original == null) {
            this.arguments = arguments
        }
    }

    /**
     * 画面のタイトル文言を設定します。
     *
     * @param title 画面のタイトル文言
     */
    fun setScreenTitle(title: String) {
        val original = this.arguments
        val arguments = original ?: Bundle()
        arguments.putString(ARGUMENT_TITLE, title)
        if (original == null) {
            this.arguments = arguments
        }
    }

    private fun loadTime() {
        val timePicker = picker
        if (timePicker != null) {
            val pickerHelper = TimePickerHelper(timePicker)
            time.hour = pickerHelper.getHour()
            time.minute = pickerHelper.getMinute()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        loadTime()
        outState.putParcelable(ARGUMENT_TIME, time)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val time: Time = if (savedInstanceState != null) {
            savedInstanceState.getParcelable(ARGUMENT_TIME)
                ?: throw RuntimeException("savedInstanceState[$ARGUMENT_TIME] is not set")
        } else {
            requireArguments().getParcelable(ARGUMENT_TIME)
                ?: throw RuntimeException("arguments[$ARGUMENT_TIME] is not set")
        }
        PreferenceLog.v(TAG) { "onCreate: time = $time" }
        this.time.hour = time.hour
        this.time.minute = time.minute
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(
            R.layout.imoya_preference_time_input_fragment, container, false
        )
        picker = view.findViewById(R.id.time)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val timePicker = picker!!
        TimePickerHelper(timePicker).setHourAndMinute(time.hour, time.minute)
        timePicker.setIs24HourView(requireArguments().getBoolean(ARGUMENT_IS_24_HOUR_VIEW))
        val title = requireArguments().getString(ARGUMENT_TITLE)
        PreferenceLog.v(TAG) { "onViewCreated: title = $title" }
        this.setTitle(title)
    }

    companion object {
        /**
         * 再生成時保存データキー:時刻
         */
        private val ARGUMENT_TIME = TimePickerFragment::class.java.name + ".CurrentTime"

        /**
         * 再生成時保存データキー:24時間表示フラグ
         */
        private val ARGUMENT_IS_24_HOUR_VIEW = TimePickerFragment::class.java.name + ".Is24HourView"

        /**
         * 再生成時保存データキー:タイトル文言
         */
        private val ARGUMENT_TITLE = TimePickerFragment::class.java.name + ".Title"

        /**
         * Tag for log
         */
        private const val TAG = "TimePickerFragment"
    }
}